codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteDistributor.java

```diff
@@ -0,0 +1,428 @@
+                childState.allocate(bytesForChild);
+                bytesAllocated += bytesForChild;
+                connectionWindow -= bytesForChild;
+            }
+            int childBytesAllocated = allocateBytesForTree(child, connectionWindow);
+            bytesAllocated += childBytesAllocated;
+            connectionWindow -= childBytesAllocated;
+            return true;
+        }
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class PriorityState {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        int unallocatedStreamableBytesForTree;
+
+        PriorityState(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        /**
+         * Recursively increments the {@link #unallocatedStreamableBytesForTree()} for this branch in
+         * the priority tree starting at the current node.
+         */
+        void unallocatedStreamableBytesForTreeChanged(int delta) {
+            unallocatedStreamableBytesForTree += delta;
+            if (!stream.isRoot()) {
+                state(stream.parent()).unallocatedStreamableBytesForTreeChanged(delta);
+            }
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored before the allocated bytes are
+            // actually written.
+            unallocatedStreamableBytesForTreeChanged(-bytes);
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocate(-allocated);
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unallocatedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unallocate all bytes.
+            resetAllocated();
+
+            // Clear the streamable bytes.
+            updateStreamableBytes(0, false);
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unallocatedStreamableBytes() {
+            return streamableBytes - allocated;
+        }
+
+        int unallocatedStreamableBytesForTree() {
+            return unallocatedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class WriteVisitor implements Http2StreamVisitor {
```

Comment:
To continue https://github.com/netty/netty/pull/4252#discussion_r40388868

@nmittler 
> @Scottmitch I'm not sure I understand.  It seems the code is effectively the same, except that you're storing the `writer` independently. Am I missing something?

I would like to avoid having to completely wrap a `Http2StreamVisitor` maintain state, have to deal with `Http2StreamVisitor`'s exception throwing and adding state and dealing with this in the control flow, and worry about re-entry, etc... My suggestion was an attempt to directly use `Http2StreamVisitor` and now worry about all of these concerns.